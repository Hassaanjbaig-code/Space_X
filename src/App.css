* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;

  --color-primary: #0290ff;
  --color-primary-dark: #0290ff;
  --color-text: #121212;
  --color-text-blue: #227fdc;
  --color-text-gray: #888;
}

.flex-center {
  display: flex;
  justify-content: center;
  align-items: center;
}

.row {
  display: flex;
}

.column {
  display: flex;
  flex-direction: column;
}

.no-style {
  list-style: none;
  text-decoration: none;
  background-color: transparent;
  border: none;
}

.seprator {
  height: 15px;
  border: 1px solid #121212;
}

.App {
  text-align: center;
}
