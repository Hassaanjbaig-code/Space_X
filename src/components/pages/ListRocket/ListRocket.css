.rocket {
  padding: 24px;
  width: 100%;
  text-align: start;
}

.rocket img {
  width: 250px;
  height: 150px;
}

.rocket-description {
  padding-left: 12px;
}

.rocket button {
  width: fit-content;
  background-color: blue;
  padding: 8px;
  color: white;
  border: none;
  border-radius: 4px;
}
