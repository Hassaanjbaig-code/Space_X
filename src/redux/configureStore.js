import { configureStore } from '@reduxjs/toolkit';
import rocketReducer from './rockets/rocketReducer';
import missionReducer from './missions/missionReducer';

export default configureStore({
  reducer: {
    rocketReducer,
    mission: missionReducer.reducer,
  },
});
