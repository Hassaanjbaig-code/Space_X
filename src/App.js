import './App.css';
import { Routes, Route } from 'react-router-dom';
import Navbar from './components/Navbar';
import ListMission from './components/pages/listMission/ListMission';
import ListRocket from './components/pages/ListRocket/ListRocket';
import Profile from './components/pages/Profile';

function App() {
  return (
    <div className="App">
      <Navbar />
      <Routes>
        <Route path="/" element={<ListRocket />} />
        <Route path="/Missions" element={<ListMission />} />
        <Route path="/myprofile" element={<Profile />} />
      </Routes>
    </div>
  );
}

export default App;
