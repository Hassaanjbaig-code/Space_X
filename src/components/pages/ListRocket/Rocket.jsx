import React from 'react'
import { useDispatch } from 'react-redux'
import { reserveRocket, unreserveRocket } from '../../../redux/rockets/rocketReducer'
import './ListRocket.css'
import PropTypes from 'prop-types';

const Rocket = (props) => {
  const dispatch = useDispatch()
  const {id, flickr_images, rocket_name, 
    description, reserved} = props.rocket

  
  return (
    <div className="row rocket">
      <img src={flickr_images} alt={`Rocket ${rocket_name}`} />
      <div className="column rocket-description">
        <h4>{rocket_name}</h4>
        <p>{reserved && (<span className='badge-reserved'>Reserved</span>)} {description}</p>
        {reserved? 
         <button onClick={()=>{dispatch(unreserveRocket(id))}}>Cancel reservation</button>
        : <button onClick={()=> {dispatch(reserveRocket(id))}}>Reserve Rocket</button>}
      </div>
    </div>
  );
};

Rocket.propType = {
  rocket: PropTypes.shape({
    id: PropTypes.string,
    flickr_images: PropTypes.string,
    rocket_name: PropTypes.string,
    description: PropTypes.string,
    reserved: PropTypes.bool,
  }),
};
export default Rocket;
