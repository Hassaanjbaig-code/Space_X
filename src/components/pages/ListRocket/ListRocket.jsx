import React, { useEffect } from 'react';
import './ListRocket.css';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import Rocket from './Rocket';
import { fetchRockets } from '../../../redux/rockets/rocketReducer';

const ListRocket = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchRockets());
  }, [dispatch]);

  const listRocket = useSelector((state) => state.rocketReducer.listRocket, shallowEqual);

  return (
    <ul>
      {(listRocket)? listRocket.map(rocket => (
        <li key={rocket.id} ><Rocket rocket={rocket}/></li>
      )): <p>Empty list</p>}
    </ul>
  );
};

export default ListRocket;
