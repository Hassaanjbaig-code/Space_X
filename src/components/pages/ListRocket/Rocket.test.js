
import renderer from 'react-test-renderer'
import '@testing-library/jest-dom'
import { Provider } from 'react-redux';
import React from 'react';
import Rocket from './Rocket';
import configureStore from '../../../redux/configureStore';
import { render, fireEvent, screen } from '@testing-library/react';
import App from '../../../App';
import ListRocket from './ListRocket';

const listRocketTest = [
  {id: '5e9d0d95eda69955f709d1eb', rocket_name: 'Falcon 1', 
  description: 'The Falcon 1 was an expendable launch system priva…launch vehicle to go into orbit around the Earth.', 
  flickr_images: 'https://imgur.com/DaCfMsj.jpg'},
  {id: '5e9d0d95eda69973a809d1ec', rocket_name: 'Falcon 9', 
  description: 'Falcon 9 is a two-stage rocket designed and manufa… satellites and the Dragon spacecraft into orbit.', 
  flickr_images: 'https://farm1.staticflickr.com/929/28787338307_3453a11a77_b.jpg'},
  {id: '5e9d0d95eda69974db09d1ed', rocket_name: 'Falcon Heavy', 
  description: 'With the ability to lift into orbit over 54 metric…hicle, the Delta IV Heavy, at one-third the cost.', 
  flickr_images: 'https://farm5.staticflickr.com/4599/38583829295_581f34dd84_b.jpg'},
  {id: '5e9d0d96eda699382d09d1ee', rocket_name: 'Starship', 
  description: 'Starship and Super Heavy Rocket represent a fully …tually replace Falcon 9, Falcon Heavy and Dragon.', 
  flickr_images: 'https://live.staticflickr.com/65535/48954138962_ee541e6755_b.jpg'}
]
    
const store = configureStore

describe('test render', () => {
  test('List Rocket should match snapshoot', () => { 
    const testRocket = {
      id: "rocket.id",
      rocket_name: "rocket.name",
      description: "rocket.description",
      flickr_images: "rocket.flickr_images[0]",
    };

    const tree = renderer.create(
      <React.StrictMode>
        <Provider store={store}>
          <Rocket rocket={testRocket} />
        </Provider>
    </React.StrictMode>,
    ).toJSON()
    expect(tree).toMatchSnapshot();
  })
})

describe('test user interaction', () => {
  test('test reserve rocket', async () => { 
    render(
      <React.StrictMode>
        <Provider store={store}>
          <ListRocket />
        </Provider>
      </React.StrictMode>,
    )

    //simulate a API REPOND to the reducer
    await store.dispatch({type: "spacex/rocket/GET_ROCKETS/fulfilled", payload: listRocketTest})

    //test the first rocket
    const title = screen.getAllByRole('heading')[0]
    const button = screen.getAllByRole('button')[0]
    
    const testRocket = listRocketTest[0]
    expect(title).toHaveTextContent(testRocket.rocket_name)
    expect(button).toHaveTextContent("Reserve Rocket")

    const description = screen.getByText(testRocket.description)

    fireEvent.click(button)
    expect(button).toHaveTextContent("Cancel reservation")

    expect(description).toHaveTextContent("Reserved "+ testRocket.description)
  })
})