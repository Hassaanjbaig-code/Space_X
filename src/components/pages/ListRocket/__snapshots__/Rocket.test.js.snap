// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test render List Rocket should match snapshoot 1`] = `
<div
  className="row rocket"
>
  <img
    alt="Rocket rocket.name"
    src="rocket.flickr_images[0]"
  />
  <div
    className="column rocket-description"
  >
    <h4>
      rocket.name
    </h4>
    <p>
       
      rocket.description
    </p>
    <button
      onClick={[Function]}
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;
